examples_tests/common/CommonAPI.h

```diff
@@ -443,128 +681,256 @@ class CommonAPI
+
+		if(graphicsQueueEnable && gpuInfo.queueFamilyProps.graphics.supportsCompute)
+			result.cpu2gpuParams.perQueue[nbl::video::IGPUObjectFromAssetConverter::EQU_COMPUTE].queue = result.queues[InitOutput<sc_image_count>::EQT_GRAPHICS];
+		else
+			result.cpu2gpuParams.perQueue[nbl::video::IGPUObjectFromAssetConverter::EQU_COMPUTE].queue = result.queues[InitOutput<sc_image_count>::EQT_COMPUTE];
+
 		return result;
 	}
-	static nbl::core::smart_refctd_ptr<nbl::video::ISwapchain> createSwapchain(uint32_t width,
+	static nbl::core::smart_refctd_ptr<nbl::video::ISwapchain> createSwapchain(
+		nbl::video::E_API_TYPE api_type,
+		uint32_t width,
 		uint32_t height,
 		uint32_t imageCount,
 		const nbl::core::smart_refctd_ptr<nbl::video::ILogicalDevice>& device,
 		const nbl::core::smart_refctd_ptr<nbl::video::ISurface>& surface,
-		nbl::video::ISurface::E_PRESENT_MODE presentMode)
+		nbl::video::ISurface::E_PRESENT_MODE requestedPresentMode,
+		nbl::video::ISurface::SFormat requestedSurfaceFormat,
+		const GPUInfo& gpuInfo)
 	{
 		using namespace nbl;
-		video::ISwapchain::SCreationParams sc_params;
+
+		nbl::asset::E_SHARING_MODE imageSharingMode;
+		if (gpuInfo.queueFamilyProps.graphics.index == gpuInfo.queueFamilyProps.present.index)
+			imageSharingMode = asset::ESM_EXCLUSIVE;
+		else
+			imageSharingMode = asset::ESM_CONCURRENT;
+		
+		nbl::video::ISurface::SFormat surfaceFormat;
+
+		if(api_type == video::EAT_VULKAN)
+		{
+			uint32_t found_format_and_colorspace = ~0u;
+			uint32_t found_format = ~0u;
+			for(uint32_t i = 0; i < gpuInfo.availableSurfaceFormats.size(); ++i)
+			{
+				const auto & supportedFormat = gpuInfo.availableSurfaceFormats[i];
+				if(requestedSurfaceFormat.format == supportedFormat.format)
+				{
+					if(found_format == ~0u)
+						found_format = i;
+					if(requestedSurfaceFormat.colorSpace.eotf == supportedFormat.colorSpace.eotf && requestedSurfaceFormat.colorSpace.primary == supportedFormat.colorSpace.primary)
+					{
+						found_format_and_colorspace = i;
+						break;
+					}
+				}
+			}
+		
+			if(found_format_and_colorspace != ~0u)
+			{
+				surfaceFormat = gpuInfo.availableSurfaceFormats[found_format_and_colorspace];
+			}
+			else if(found_format != ~0u) // fallback
+			{
+				assert(false && "Fallback: requested 'colorspace' is not supported.");
+				surfaceFormat = gpuInfo.availableSurfaceFormats[found_format];
+			}
+			else
+			{
+				assert(false && "Fallback: requested 'format' and 'colorspace' is not supported.");
+				surfaceFormat = gpuInfo.availableSurfaceFormats[0];
+			}
+
+			bool presentModeSupported = (gpuInfo.availablePresentModes & requestedPresentMode) != 0;
+			assert(presentModeSupported);
+			if(!presentModeSupported) // fallback 
+			{
+				requestedPresentMode = nbl::video::ISurface::E_PRESENT_MODE::EPM_FIFO;
+				assert(false && "Fallback: requested 'present mode' is not supported");
+			}
+		}
```

Comment:
@achalpandeyy what happened to making OpenGL report its surface formats and present modes?

Reply:
It reports present modes correctly. Need to do surface formats.